### ISSUE TITLE
Incremental compile incorrectly rebuilds in reactor project

### ISSUE THREAD

**Issue by timw** [davidB/scala-maven-plugin#368]
I've got a Maven reactor (e.g. multiple module) project, with inter-module dependencies, and observed that repeated incremental compiles sometimes (e.g. `maven compile`) trigger a rebuild when no source files have changed.

This could well be a Zinc problem (I tend to think it is based on my debugging), but raising it here first in case I'm missing something in how `scala-maven-plugin` drives the incremental compiler.

What I've observed is:
- In a (minimally) two module project, one of which depends on the other
- In the depending project, a class in an object is accessed
- Repeated compiles of the two projects with no source changes (e.g. a full reactor build, or a partial build with `-pl` and `-am`) will result in some of the source files in the depending project being re-compiled.
- The incorrect rebuild is not entirely deterministic - i.e. in a minimal 1 source file per module test case, repeated compiles will sometimes compile nothing, and sometimes (more often than not) compile the source file.
- If you perform a single module compile (non reactor) from the module directory, the first compile will recompile source files, and subsequent compiles will skip all files. i.e. in this case, the dependencies are coming from an installed `.jar`, and the incremental compiler works as expected.

I've uploaded a minimal project reproducing the issue at https://github.com/timw/maven-scala-incremental-confusion

From what I can debug so far, it looks like the previous build analysis used by Zinc is incorrect when it's stored:
- In the cached analysis file, the object type (e.g. `package$ObjectType$.class`) is registered as coming from the binary dependency `/src/folder/path/package/$ObjectType.class` (note the missing trailing `$`).
- When the incremental compiler tries to detect changes, it (correctly) calculates the `package$ObjectType$.class` type should be in `/src/folder/path/package/$ObjectType$.class` and (incorrectly) determines that the binary dependency containing the type has changed, triggering a recompile of source files that use that type.
- I have no concrete theory to explain why this isn't entirely deterministic, apart from hand-waving about timestamps and possible filesystem interactions.

Dev system here is OS X, Java 11, Maven 3.6.1, but I've seen the incremental recompile in Docker/Linux builds as well.

**Comment by slandelle** [comment]
It looks like a Zinc bug to me.
And indeed, behavior in non deterministic.

Just run `mvn compile -X` multiple times (so no jar generation involved) and you'll see module2 being sometimes recompiled with:

```
[DEBUG] Invalidating '/Users/slandelle/Downloads/maven-scala-incremental-confusion-master/module1/target/classes/package1/Producer.class' because package1.Producer$Domain is now provided by /Users/slandelle/Downloads/maven-scala-incremental-confusion-master/module1/target/classes/package1/Producer$Domain.class
[DEBUG]
Initial source changes:
	removed:Set()
	added: Set()
	modified: Set()
Invalidated products: Set()
External API changes: API Changes: Set()
Modified binary dependencies: Set(/Users/slandelle/Downloads/maven-scala-incremental-confusion-master/module1/target/classes/package1/Producer.class)
Initial directly invalidated classes: Set()

Sources indirectly invalidated by:
	product: Set()
	binary dep: Set(/Users/slandelle/Downloads/maven-scala-incremental-confusion-master/module2/src/main/scala/package2/Consumer.scala)
	external source: Set()
[DEBUG] All sources are invalidated.
[DEBUG] Initial set of included nodes:
[DEBUG] Recompiling all sources: number of invalidated sources > 50.0% of all sources
[INFO] Compiling 1 Scala source to /Users/slandelle/Downloads/maven-scala-incremental-confusion-master/module2/target/classes ..
```

**Comment by slandelle** [comment]
@eed3si9n Could you please share some light?

**Comment by eed3si9n** [comment]
Without digging into the details I can't tell for sure but it's possible that there's a Zinc bug or somehow the Analysis file association is missing for the subproject. Can it be reproduced using sbt?

**Comment by timw** [comment]
I've updated the sample project with a basic sbt multi-project build using sbt 1.3.3 (which uses the same `zinc-compile:1.3.1` dependency.
Can't reproduce with repeated `sbt compile` invocations.

**Comment by timw** [comment]
Debugged this in a bit more detail.

When the analysis is initially computed and stored, a binary dependency on: `class Domain` qualified by `class Product` is calculated (this is in the compiler bridge).
This results in a binary dependency for `Product$Domain` pointing to the `Product.class` file (this is using `scala.lang.ClassSymbol.associatedFile`, so I assume that's the canonical position.
Strictly the class is actually defined in a different class file, but I guess for incremental compile purposes the `Product.class` file would be represent the containing compilation unit and have the same compilation lifecycle.)

When `sbt.internal.inc.IncrementalCommon#isBinaryModified` is invoked on a subsequent compilation to check the class, and subsequently in `compareOriginClassFile` `sbt.internal.inc.Locate#resolve` is used to find the class file for the class.
The problem is that `Locate#resolve` returns the actual class file the class is defined in, and not the associated class file the `ClassSymbol#associatedFile` returned in the analysis calculation, resulting in an invalidation of the dependency and any source files depending on it.

**Comment by timw** [comment]
A bit more debugging...

This does look like a Zinc bug, but sbt sidesteps it by wiring in the analysis for dependency source projects (which scala-maven-plugin doesn't do at the moment).
Essentially this means implementing `xsbti.compile.PerClasspathEntryLookup#analysis` to locate the analysis file for a particular classpath entry, which allows Zinc to resolve items like `Product$Domain.class` to a source file in the dependency project.
The dependency is then added as a external source dependency, not an external binary dependency and sidesteps the bug.

It seems possible to walk the project references for the current project, pick up the cached analysis files and produce this map, which should sort things out.
I'm not sure what the best way to locate the `scala-maven-plugin` config from other projects is though - i.e. if profiles/executions/non default `analysisCacheFile` are being used.

**Comment by slandelle** [comment]
Great findings, indeed implementing PerClasspathEntryLookup#analysis does the trick.

I agree it would be very complicated to cover anything but the default configuration and maven project layout. Typically, we would have to assume output directory is `target`. 
Still, I think it's worth it.

I'll come up with a PR, except if anyone wants to take a shot at it.

**Comment by slandelle** [comment]
Fixed!

@davidB Could you please release?

**Comment by timw** [comment]
Thanks @slandelle - Zinc only calls `PerClasspathEntryLookup#analysis` once per classpath entry (caches it for the lifetime of the incremental compiler) so this fix looks good for now.

**Comment by davidB** [comment]
fix included in 4.3.0
